ASE: Use `guess=true` while checking parser support

rashidrafeek opened the pull request:
Is there a reason for setting `guess=false` while checking parser support? Only if `guess=true` will ASE guess the extension based on extension and therefore can read many formats like VASP POSCARS (\*.vasp, POSCAR\*), QE input (\*.in) etc. Currently, it shows an error: `Python: UnknownFileTypeError: Could not guess file type` due to setting `guess=false`.

rashidrafeek committed:
ASE: Use `guess=true` while checking parser support

mfherbst commented:
Thanks for the suggestion.

Hmm I guess that was to be careful and to avoid issues when file extensions are not unique. For sure we should make this a parameter exposed to the user (eg a flag in ASEParser, but before changing it we should double check this has no negative side effects.

We could also have two ase parser objects in the default list of parsers: the first more strict and the second laxer. This would make it more obvious what's the cause in case sth. breaks in the future.

mfherbst commented:
Oh also please provide example files for edge cases / unexpected behaviour if you can. Always good to add them to the atomsio test suite, because the whole parsing stuff is indeed quite messy.

rashidrafeek commented:
Sorry for the late reply. I was a little busy and could not respond to you comments.

I think defining these parameters in the `AseParser` object with a default value using `@kwdef` would be better instead of defining multiple parser objects. For example, we can redefine `AseParser` as,
```
@kwdef struct AseParser <: AbstractParser 
    read::Bool = false
    guess::Bool = false
end

function AtomsIO.supports_parsing(parser::AseParser, file; save, trajectory)
    ...
        format = ase.io.formats.filetype(file; read=parser.read, guess=parser.guess)
    ...
end
```
Additionally, it seems that currently the `read` argument above (which enables reading the file to determine filetype) is mapped to the `save` argument of `supports_parsing`. AFAIU, these two flags are independent and mapping them leads to incorrect behavior. For example, even though ASE can both read and write "*.xsf" files, currently it gives an error when `save=false`:
```
julia> using AtomsIO: supports_parsing
       using AtomsIOPython: AseParser

julia> supports_parsing(AseParser(), "test.xsf"; save=true, trajectory=false)
true

julia> supports_parsing(AseParser(), "test.xsf"; save=false, trajectory=false)
ERROR: UndefVarError: `PyException` not defined
Stacktrace:
 [1] supports_parsing(::AseParser, file::String; save::Bool, trajectory::Bool)
   @ AtomsIOPython ~/GitProjects/AtomsIO/lib/AtomsIOPython/src/ase.jl:22
 [2] top-level scope
   @ REPL[46]:1

caused by: Python: UnknownFileTypeError: Could not guess file type
...
```
This behavior is present for "*.vasp" files also. 

There are individual functions `can_read` and `can_write` in the `IOFormat` object of ASE to determine whether ASE supports reading/writing for a specific format. I think these should be used instead. To change this, the following lines:
https://github.com/mfherbst/AtomsIO.jl/blob/f15cec642a2478546af93b5177b4b6116acd90cb/lib/AtomsIOPython/src/ase.jl#L32-L34
 would need to changed to
```
    ioformat = ase.io.formats.ioformats[format]
    elseif save
        return ioformat.can_write
    else
        return ioformat.can_read
    end
```

If these changes seem fine, I will modify the PR accordingly. Let me know what you think.

mfherbst commented:
Sounds good. I'm a bit confused what the `read` ASE kwarg is then actually for. Is there no way we can give it a sane default and not worry about it from then on?

rashidrafeek committed:
Define AseParser() with @kwdef, fix check for format read and write

- Add arguments to `filetype` function in `AseParser` struct with
  default same as ASE.
- Check whether ASE can read or write format was not done. Add these
  checks using `can_read` and `can_write` functios of `ioformat`.

mfherbst commented on lib/AtomsIOPython/src/ase.jl:
```diff
@@ -11,26 +11,48 @@ Supported formats:
+    end
+
+    ioformat = ase.io.formats.ioformats[format]
+    supports_trajectory = '+' in ase.io.formats.ioformats[format].code
```
'+' in ioformat.code

mfherbst commented on lib/AtomsIOPython/src/ase.jl:
```diff
@@ -11,26 +11,48 @@ Supported formats:
+    supports_trajectory = '+' in ase.io.formats.ioformats[format].code
+
+    if save # Check whether ASE can write format
+        if Bool(ioformat.can_write)
```
pyconvert(Bool, ioformat.can_write)

rashidrafeek replied:
That would require making PythonCall a dependency, which is why I did this. Is that fine? Currently ASEconvert is used to import ASE.

mfherbst replied:
I think ASEconvert export pyconvert.

mfherbst replied:
Yes it does:
https://github.com/mfherbst/ASEconvert.jl/blob/master/src/ASEconvert.jl#L10

rashidrafeek replied:
Oh I didn't check that. Thanks. Will change it.

mfherbst commented on lib/AtomsIOPython/src/ase.jl:
```diff
@@ -11,26 +11,48 @@ Supported formats:
+    ioformat = ase.io.formats.ioformats[format]
+    supports_trajectory = '+' in ase.io.formats.ioformats[format].code
+
+    if save # Check whether ASE can write format
```
I would structure it as

If save && can write
   Return true if trajectory matches
Elseif !save && can read
   ...
End
Return false

mfherbst reviewed (commented):
Thanks getting close. Just a small refactor ;)

mfherbst commented on lib/AtomsIOPython/src/ase.jl:
```diff
@@ -11,26 +11,48 @@ Supported formats:
     format = ""
     try
-        format = ase.io.formats.filetype(file; read=!save, guess=false)
+        format = ase.io.formats.filetype(file; read=parser.read, guess=parser.guess)
```
I don't think this is correct, since `ase.io.filetype` will try to open the file if `read=true`. If the file does not exist (because you want to save it) this will not work. I think the original `!save` is correct here.

```suggestion
        # read=true causes ASE to open the file, read a few bytes and check for magic bytes
        format = ase.io.formats.filetype(file; read=!save, guess=parser.guess)
```

rashidrafeek replied:
Yeah this is true. Due to this the tests are not passing currently. I had not thought about this earlier.

mfherbst replied:
That's why we have tests :smile:

mfherbst commented on lib/AtomsIOPython/src/ase.jl:
```diff
@@ -11,26 +11,48 @@ Supported formats:
 """
-struct AseParser <: AbstractParser end
+@kwdef struct AseParser <: AbstractParser 
+    read::Bool = true
```
```suggestion
```
(see below)

rashidrafeek replied:
Yeah, this is an option. But then it won't be possible to set `read=false` when loading files. It would be useful to make it customizable.

mfherbst replied:
True, but I find it hard to imagine to have a use case where I wanted to read a file, but not want to open it first to read the magic bytes in order to determine the file type. Do you have one in mind?

rashidrafeek replied:
The `ase.io.formats.filetype` functions seems to have an unexpected behaviour when `read=True` and `guess=False` (which is the default behavior while reading a file in AtomsIO master branch). For example,
```
>>> from ase.io import formats
>>> formats.filetype("test.xsf", read=True, guess=True)
'xsf'
>>> formats.filetype("test.xsf", read=False, guess=False)
'xsf'
>>> formats.filetype("test.xsf", read=False, guess=True)
'xsf'
>>> formats.filetype("test.xsf", read=True, guess=False)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/rashid/Apps/mambaforge/envs/mlff/lib/python3.9/site-packages/ase/io/formats.py", line 954, in filetype
    raise UnknownFileTypeError('Could not guess file type')
ase.io.formats.UnknownFileTypeError: Could not guess file type
```
This occurs for other formats also and this is the original reason I came up with this issue. I think this is actually a bug in the implementation of `ase.io.formats.filetype` function in ASE. But I thought having it be customizable in AtomsIO would be better.

mfherbst replied:
Yikes, in case not yet done, please open an issue with ASE. Yes in this case let's have `guess = true` the default and make a comment in the code.

rashidrafeek replied:
Submitted an issue: https://gitlab.com/ase/ase/-/issues/1378

rashidrafeek committed:
`read` should not be set while writing, remove `read` from `AseParser`

Co-authored-by: Michael F. Herbst <info@michael-herbst.com>

rashidrafeek committed:
Implement code review suggestions

rashidrafeek committed:
Add xsf format to supports parsing tests

rashidrafeek commented:
I have implemented your suggestions. And have removed the `read` argument from the `AseParser` for now as you suggested. The issue with the `read` argument would not occur if `guess=true` is made the default. Wouldn't this default make more sense as this is the current default in ASE? 

I have also added a test for xsf format which would not pass in the current master. Let me know if these changes are okay.

mfherbst committed:
Slightly more clear

mfherbst committed:
fixes

mfherbst committed:
bump version

mfherbst committed:
add warning

mfherbst committed:
Test for warning

mfherbst commented:
@rashidrafeek Could you add tests in `lib/AtomsIOPython/test/ase.jl` for the cases you would want to have supported, just to ensure that we don't break things in the future ?

mfherbst committed:
ups

mfherbst committed:
Fix test

mfherbst committed:
fix

rashidrafeek committed:
ASE: Add tests for xyz , vasp. Fix mistake in warning text

rashidrafeek commented:
Sorry for the delay here. I have added .xyz and .vasp to the tests, which I use regularly. 

Also, there was a mistake in the wording of the warning. The bug occurs when trying to read, not write (i.e, save=false). Hope it is good to go now.

mfherbst commented:
Thanks!
